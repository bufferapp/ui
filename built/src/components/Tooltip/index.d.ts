export { default } from './Tooltip';
//# sourceMappingURL=index.d.ts.map