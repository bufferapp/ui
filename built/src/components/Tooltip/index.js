export { default } from './Tooltip';
//# sourceMappingURL=index.js.map