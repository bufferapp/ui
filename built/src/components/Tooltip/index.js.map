{"version":3,"file":"index.js","sourceRoot":"","sources":["../../../../src/components/Tooltip/index.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAE,MAAM,WAAW,CAAA","sourcesContent":["export { default } from './Tooltip'\n"]}