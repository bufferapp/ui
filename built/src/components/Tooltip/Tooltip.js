import React from 'react';
import PropTypes from 'prop-types';
import '@reach/tooltip/styles.css';
import * as Styles from './style';
import { gray, white } from '../style/colors';
class Tooltip extends React.Component {
    constructor(props) {
        super(props);
        this.renderLabel = (label, hotkey, customLabel = null) => (React.createElement(Styles.LabelWrapper, null,
            label && (React.createElement(Styles.Label, { color: white },
                React.createElement(Styles.HotkeyWrapper, null,
                    label,
                    hotkey && (React.createElement(Styles.Label, { color: gray, isHotkey: true }, hotkey))))),
            customLabel));
        this.setTooltipPosition = this.setTooltipPosition.bind(this);
        this.state = {
            childWidth: 0,
        };
    }
    componentDidMount() {
        this.setTooltipPosition();
        window.addEventListener('resize', this.setTooltipPosition);
    }
    componentWillUnmount() {
        window.removeEventListener('resize', this.setTooltipPosition);
    }
    setTooltipPosition() {
        if (this.tooltipWrapper) {
            const childWidth = this.tooltipWrapper.children[0].children[0].getBoundingClientRect()
                .width;
            this.setState({
                childWidth,
            });
        }
    }
    getTooltipPosition(triggerRect, tooltipRect, position, verticalAlign) {
        const { childWidth } = this.state;
        const gap = 8;
        const triggerCenter = triggerRect.left + childWidth / 2;
        const left = triggerCenter - tooltipRect.width / 2;
        const maxLeft = window.innerWidth - tooltipRect.width - 2;
        const verticalCenter = triggerRect.top +
            triggerRect.height -
            (triggerRect.height + tooltipRect.height) / 2 -
            window.scrollY;
        const verticalTop = triggerRect.top;
        const topPosition = verticalAlign === 'top' ? verticalTop : verticalCenter;
        const newPosition = {
            left: Math.min(Math.max(2, left), maxLeft) + window.scrollX,
            top: triggerRect.bottom + gap + window.scrollY,
        };
        switch (position) {
            case 'top':
                newPosition.top =
                    triggerRect.top - tooltipRect.height - gap - window.scrollY;
                break;
            case 'right':
                newPosition.left = triggerRect.left + childWidth + gap + window.scrollX;
                newPosition.top = topPosition;
                break;
            case 'left':
                newPosition.left =
                    triggerRect.left - tooltipRect.width - gap - window.scrollX;
                newPosition.top = topPosition;
                break;
            default:
        }
        return newPosition;
    }
    render() {
        const { children, label, position, verticalAlign, hotkey, customLabel, opacity, } = this.props;
        const renderTooltip = label || customLabel;
        return (React.createElement(React.Fragment, null, renderTooltip ? (React.createElement(Styles.TooltipWrapper, { ref: (node) => (this.tooltipWrapper = node) },
            React.createElement(Styles.TooltipStyled, { label: this.renderLabel(label, hotkey, customLabel), position: (triggerRect, tooltipRect) => this.getTooltipPosition(triggerRect, tooltipRect, position, verticalAlign), style: Styles.TooltipStyle, opacity: opacity },
                React.createElement("div", null, children)))) : (React.createElement("div", null, children))));
    }
}
Tooltip.propTypes = {
    children: PropTypes.node.isRequired,
    label: PropTypes.string,
    position: PropTypes.string,
    verticalAlign: PropTypes.string,
    hotkey: PropTypes.string,
    customLabel: PropTypes.node,
    opacity: PropTypes.number,
};
Tooltip.defaultProps = {
    label: '',
    position: 'bottom',
    verticalAlign: 'center',
    hotkey: '',
    customLabel: '',
    opacity: 1,
};
export default Tooltip;
//# sourceMappingURL=Tooltip.js.map