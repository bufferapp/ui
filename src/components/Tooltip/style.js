import styled from 'styled-components';
import Tooltip from '@reach/tooltip';
import {
  fontFamily,
  fontSizeSmall,
  fontWeight,
  lineHeightSmall
} from '../style/fonts';
import {
  grayDarker,
  white
} from '../style/colors';

export const TooltipWrapper = styled.div`
`;

export const TooltipStyled = styled(Tooltip)`
  opacity: ${props => props.opacity}
`;

export const TooltipStyle = {
  position: "absolute",
  background: grayDarker,
  color: white,
  border: "none",
  borderRadius: "4px",
  padding: "8px",
  maxWidth: "200px",
  whiteSpace: "normal",
  zIndex: 9999,
}

export const Label = styled.label`
  color: ${props => props.color};
  font-family: ${fontFamily};
  font-size: ${props => props.isHotkey ? '10px' : fontSizeSmall};
  margin-left: ${props => props.isHotkey ? '8px' : '0'};
  font-weight: ${fontWeight};
  line-height: ${lineHeightSmall};
  cursor: inherit;
`;

export const HotkeyWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
`;

export const LabelWrapper = styled.div`
`;
