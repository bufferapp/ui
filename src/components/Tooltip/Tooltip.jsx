import React from 'react';
import PropTypes from 'prop-types';
import '@reach/tooltip/styles.css';
import * as Styles from './style';
import {
  gray,
  white
} from '../style/colors';

class Tooltip extends React.Component {
  constructor(props) {
    super(props);

    this.setTooltipPosition = this.setTooltipPosition.bind(this);
    this.state = {
      childWidth: 0,
    };
  }

  componentDidMount() {
    this.setTooltipPosition();
    window.addEventListener("resize", this.setTooltipPosition);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.setTooltipPosition);
  }

  setTooltipPosition() {
    // Getting the first child width to calculate Tooltip position
    if (this.tooltipWrapper) {
      const childWidth = this.tooltipWrapper.children[0].children[0].getBoundingClientRect().width;
      this.setState({
        childWidth,
      });
    }
  }

  /**
   * Adjusting the styles according to the desired position
   * The tooltip should be vertically or horizontally centered
   */
  getTooltipPosition(triggerRect, tooltipRect, position, verticalAlign) {
    const { childWidth } = this.state;
    const gap = 8;
    const triggerCenter = triggerRect.left + childWidth / 2;
    const left = triggerCenter - tooltipRect.width / 2;
    const maxLeft = window.innerWidth - tooltipRect.width - 2;
    const verticalCenter =
      triggerRect.top +
      triggerRect.height -
      (triggerRect.height + tooltipRect.height) / 2 -
      window.scrollY;
    const verticalTop = triggerRect.top;
    const topPosition =
      verticalAlign === 'top' ? verticalTop : verticalCenter;

    const newPosition = {
      left: Math.min(Math.max(2, left), maxLeft) + window.scrollX,
      top: triggerRect.bottom + gap + window.scrollY,
    };

    switch (position) {
      case 'top':
        newPosition.top = triggerRect.top - tooltipRect.height - gap - window.scrollY;
        break;

      case 'right':
        newPosition.left = triggerRect.left + childWidth + gap + window.scrollX;
        newPosition.top = topPosition;
        break;

      case 'left':
        newPosition.left = triggerRect.left - tooltipRect.width - gap - window.scrollX;
        newPosition.top = topPosition;
        break;

      default:
    }

    return newPosition;
  }

  /**
   * Rendering label with hotkey option if available
   */
  renderLabel = (label, hotkey, customLabel = null) => (
    <Styles.LabelWrapper>
      {label && (
        <Styles.Label
          color={white}
        >
          <Styles.HotkeyWrapper>
            {label}
            {hotkey &&
              <Styles.Label color={gray} isHotkey>{hotkey}</Styles.Label>}
          </Styles.HotkeyWrapper>
        </Styles.Label>
      )}
      {customLabel}
    </Styles.LabelWrapper>
  );

  render() {
    const { children, label, position, verticalAlign, hotkey, customLabel, opacity } = this.props;

    const renderTooltip = label || customLabel;

    // @todo: remove style from here and use the Styled component
    // We are currently adding the stylings with the style tag,
    // instead of the adding it in the Styled component
    // because we still need to figure out a way to load the css file
    // properly, and being able to use our customs styles.
    return (
      <React.Fragment>
        {renderTooltip ? (
          <Styles.TooltipWrapper ref={node => this.tooltipWrapper = node}>
            <Styles.TooltipStyled
              label={this.renderLabel(label, hotkey, customLabel)}
              position={(triggerRect, tooltipRect) => this.getTooltipPosition(triggerRect, tooltipRect, position, verticalAlign)}
              style={Styles.TooltipStyle}
              opacity={opacity}
            >
              <div>
                {children}
              </div>
            </Styles.TooltipStyled>
          </Styles.TooltipWrapper>
        ) : (
          <div>
            {children}
          </div>
        )}
      </React.Fragment>
    )
  }
}

Tooltip.propTypes = {
  /** The component being wrapped */
  children: PropTypes.node.isRequired,

  /** The tooltip label */
  label: PropTypes.string,

  /** The tooltip position */
  position: PropTypes.string,

  /** The tooltip vertical position: this only applies for left and right positioned tooltip */
  verticalAlign: PropTypes.string,

  /** The tooltip position */
  hotkey: PropTypes.string,

  /** Custom Label */
  customLabel: PropTypes.node,

  /** Custom Opacity */
  opacity: PropTypes.number,
};

Tooltip.defaultProps = {
  label: '',
  position: 'bottom',
  verticalAlign: 'center',
  hotkey: '',
  customLabel: '',
  opacity: 1
};

export default Tooltip;
