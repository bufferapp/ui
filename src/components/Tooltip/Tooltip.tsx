import React from 'react'
import PropTypes from 'prop-types'
import '@reach/tooltip/styles.css'
import * as Styles from './style'
import { gray, white } from '../style/colors'

class Tooltip extends React.Component {
  tooltipWrapper: any = null

  // @ts-expect-error TS(7006) FIXME: Parameter 'props' implicitly has an 'any' type.
  constructor(props) {
    super(props)

    this.setTooltipPosition = this.setTooltipPosition.bind(this)
    this.state = {
      childWidth: 0,
    }
  }

  componentDidMount() {
    this.setTooltipPosition()
    window.addEventListener('resize', this.setTooltipPosition)
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.setTooltipPosition)
  }

  setTooltipPosition() {
    // Getting the first child width to calculate Tooltip position
    if (this.tooltipWrapper) {
      const childWidth =
        this.tooltipWrapper.children[0].children[0].getBoundingClientRect()
          .width
      this.setState({
        childWidth,
      })
    }
  }

  /**
   * Adjusting the styles according to the desired position
   * The tooltip should be vertically or horizontally centered
   */
  // @ts-expect-error TS(7006) FIXME: Parameter 'triggerRect' implicitly has an 'any' ty... Remove this comment to see the full error message
  getTooltipPosition(triggerRect, tooltipRect, position, verticalAlign) {
    // @ts-expect-error TS(2339) FIXME: Property 'childWidth' does not exist on type 'Read... Remove this comment to see the full error message
    const { childWidth } = this.state
    const gap = 8
    const triggerCenter = triggerRect.left + childWidth / 2
    const left = triggerCenter - tooltipRect.width / 2
    const maxLeft = window.innerWidth - tooltipRect.width - 2
    const verticalCenter =
      triggerRect.top +
      triggerRect.height -
      (triggerRect.height + tooltipRect.height) / 2 -
      window.scrollY
    const verticalTop = triggerRect.top
    const topPosition = verticalAlign === 'top' ? verticalTop : verticalCenter

    const newPosition = {
      left: Math.min(Math.max(2, left), maxLeft) + window.scrollX,
      top: triggerRect.bottom + gap + window.scrollY,
    }

    switch (position) {
      case 'top':
        newPosition.top =
          triggerRect.top - tooltipRect.height - gap - window.scrollY
        break

      case 'right':
        newPosition.left = triggerRect.left + childWidth + gap + window.scrollX
        newPosition.top = topPosition
        break

      case 'left':
        newPosition.left =
          triggerRect.left - tooltipRect.width - gap - window.scrollX
        newPosition.top = topPosition
        break

      default:
    }

    return newPosition
  }

  /**
   * Rendering label with hotkey option if available
   */
  // @ts-expect-error TS(7006) FIXME: Parameter 'label' implicitly has an 'any' type.
  renderLabel = (label, hotkey, customLabel = null) => (
    <Styles.LabelWrapper>
      {label && (
        <Styles.Label color={white}>
          <Styles.HotkeyWrapper>
            {label}
            {hotkey && (
              <Styles.Label color={gray} isHotkey>
                {hotkey}
              </Styles.Label>
            )}
          </Styles.HotkeyWrapper>
        </Styles.Label>
      )}
      {customLabel}
    </Styles.LabelWrapper>
  )

  render() {
    const {
      children,
      // @ts-expect-error TS(2339) FIXME: Property 'label' does not exist on type 'Readonly<... Remove this comment to see the full error message
      label,
      // @ts-expect-error TS(2339) FIXME: Property 'position' does not exist on type 'Readon... Remove this comment to see the full error message
      position,
      // @ts-expect-error TS(2339) FIXME: Property 'verticalAlign' does not exist on type 'R... Remove this comment to see the full error message
      verticalAlign,
      // @ts-expect-error TS(2339) FIXME: Property 'hotkey' does not exist on type 'Readonly... Remove this comment to see the full error message
      hotkey,
      // @ts-expect-error TS(2339) FIXME: Property 'customLabel' does not exist on type 'Rea... Remove this comment to see the full error message
      customLabel,
      // @ts-expect-error TS(2339) FIXME: Property 'opacity' does not exist on type 'Readonl... Remove this comment to see the full error message
      opacity,
    } = this.props

    const renderTooltip = label || customLabel

    // @todo: remove style from here and use the Styled component
    // We are currently adding the stylings with the style tag,
    // instead of the adding it in the Styled component
    // because we still need to figure out a way to load the css file
    // properly, and being able to use our customs styles.
    return (
      <React.Fragment>
        {renderTooltip ? (
          <Styles.TooltipWrapper ref={(node) => (this.tooltipWrapper = node)}>
            <Styles.TooltipStyled
              label={this.renderLabel(label, hotkey, customLabel)}
              // @ts-expect-error TS(7006) FIXME: Parameter 'triggerRect' implicitly has an 'any' ty... Remove this comment to see the full error message
              position={(triggerRect, tooltipRect) =>
                this.getTooltipPosition(
                  triggerRect,
                  tooltipRect,
                  position,
                  verticalAlign,
                )
              }
              style={Styles.TooltipStyle}
              opacity={opacity}
            >
              <div>{children}</div>
            </Styles.TooltipStyled>
          </Styles.TooltipWrapper>
        ) : (
          <div>{children}</div>
        )}
      </React.Fragment>
    )
  }
}

// @ts-expect-error TS(2339) FIXME: Property 'propTypes' does not exist on type 'typeo... Remove this comment to see the full error message
Tooltip.propTypes = {
  /** The component being wrapped */
  children: PropTypes.node.isRequired,

  /** The tooltip label */
  label: PropTypes.string,

  /** The tooltip position */
  position: PropTypes.string,

  /** The tooltip vertical position: this only applies for left and right positioned tooltip */
  verticalAlign: PropTypes.string,

  /** The tooltip position */
  hotkey: PropTypes.string,

  /** Custom Label */
  customLabel: PropTypes.node,

  /** Custom Opacity */
  opacity: PropTypes.number,
}

// @ts-expect-error TS(2339) FIXME: Property 'defaultProps' does not exist on type 'ty... Remove this comment to see the full error message
Tooltip.defaultProps = {
  label: '',
  position: 'bottom',
  verticalAlign: 'center',
  hotkey: '',
  customLabel: '',
  opacity: 1,
}

export default Tooltip
