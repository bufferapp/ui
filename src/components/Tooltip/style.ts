import styled from 'styled-components'
import Tooltip from '@reach/tooltip'
import {
  fontFamily,
  fontSizeSmall,
  fontWeight,
  lineHeightSmall,
} from '../style/fonts'
import { grayDarker, white } from '../style/colors'

export const TooltipWrapper = styled.div``

export const TooltipStyled: any = styled(Tooltip)`
  opacity: ${(props): string => props.opacity};
`

export const TooltipStyle = {
  position: 'absolute',
  background: grayDarker,
  color: white,
  border: 'none',
  borderRadius: '4px',
  padding: '8px',
  maxWidth: '200px',
  whiteSpace: 'normal',
  zIndex: 9999,
}

export const Label = styled.label<{ color: string; isHotkey?: boolean }>`
  color: ${(props): string => props.color};
  font-family: ${fontFamily};
  font-size: ${(props): string => (props.isHotkey ? '10px' : fontSizeSmall)};
  margin-left: ${(props): string => (props.isHotkey ? '8px' : '0')};
  font-weight: ${fontWeight};
  line-height: ${lineHeightSmall};
  cursor: inherit;
`

export const HotkeyWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
`

export const LabelWrapper = styled.div``
