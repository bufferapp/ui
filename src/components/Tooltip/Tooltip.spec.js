/* eslint-disable react/jsx-filename-extension */
import snap from 'jest-auto-snapshots'
import 'jest-styled-components'

import Tooltip from './Tooltip'

snap(Tooltip, './Tooltip.jsx')
