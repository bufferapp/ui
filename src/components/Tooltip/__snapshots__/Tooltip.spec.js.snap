// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-auto-snapshots > Tooltip Matches snapshot when passed all props 1`] = `
<React.Fragment>
  <ForwardRef(styled.div)>
    <ForwardRef(Styled(Tooltip))
      label={
        <ForwardRef(styled.div)>
          <ForwardRef(styled.label)
            color="#FFFFFF"
          >
            <ForwardRef(styled.div)>
              jest-auto-snapshots String Fixture
              <ForwardRef(styled.label)
                color="#B8B8B8"
                isHotkey={true}
              >
                jest-auto-snapshots String Fixture
              </ForwardRef(styled.label)>
            </ForwardRef(styled.div)>
          </ForwardRef(styled.label)>
          <NodeFixture />
        </ForwardRef(styled.div)>
      }
      opacity={1}
      position={[Function]}
      style={
        Object {
          "background": "#3D3D3D",
          "border": "none",
          "borderRadius": "4px",
          "color": "#FFFFFF",
          "maxWidth": "200px",
          "padding": "8px",
          "position": "absolute",
          "whiteSpace": "normal",
          "zIndex": 9999,
        }
      }
    >
      <div>
        <NodeFixture />
      </div>
    </ForwardRef(Styled(Tooltip))>
  </ForwardRef(styled.div)>
</React.Fragment>
`;

exports[`jest-auto-snapshots > Tooltip Matches snapshot when passed only required props 1`] = `
<React.Fragment>
  <div>
    <NodeFixture />
  </div>
</React.Fragment>
`;
